### PR TITLE
Introduce syntax of looping over arrays…

### PR DESCRIPTION
**Pull request by woocash2**
…as a syntax sugar of the regular loop.

Added e2e test for array loop.

### CODE REVIEW

File: app/src/main/kotlin/compiler/syntax/LanguageTokens.kt [hunk 1]
```diff
@@ -39,6 +39,8 @@ object LanguageTokens {
             // Array related keywords
             Pair(TokenType.LENGTH, "długość"),
             Pair(TokenType.ARRAY_ALLOCATION, "ciąg"),
+            Pair(TokenType.LOOP, "dla"),
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/syntax/LanguageTokens.kt:42]: I think `ARRAY_FOR_EACH` would be more descriptive.

File: app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt [hunk 2]
```diff
@@ -314,4 +314,18 @@ class CorrectProgramsE2eTest {
             """
         )
     }
+
+    @Test
+    fun `test looping over arrays`() {
```
Comment (tomekb234) [inline app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt:319]: Also at least one integration test (in the `tests/` folder) would be useful.

File: app/src/main/kotlin/compiler/syntax/AstFactory.kt [hunk 3]
```diff
@@ -19,6 +19,8 @@ import compiler.syntax.utils.TokenRegexParser
 class AstFactory(private val diagnostics: Diagnostics) {
     class AstCreationFailed : CompilationFailed()
 
+    // We maintain the following counter to generate unique names of variables introduced by the associated syntactic sugar.
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/syntax/AstFactory.kt:22]: This comment seems to be outdated.

**Review by tomekb234 — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/compiler/syntax/AstFactory.kt [hunk 4]
```diff
@@ -409,6 +411,60 @@ class AstFactory(private val diagnostics: Diagnostics) {
                 val bodyBlock = processMaybeBlock(children[4])
                 Statement.Loop(conditionExpr, bodyBlock, combineLocations(children))
             }
+            in listOf(Productions.nonBraceStatementArrayLoop, Productions.nonIfNonBraceStatementArrayLoop) -> {
+                val arrayExpression = processExpression(children[6])
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/syntax/AstFactory.kt:415]: Please extract this code into a method (e.g. `makeArrayLoop`) which could also be used in #179.

File: app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt (hunk 2 above)
Comment (woocash2) [reply app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt:319]: You mean in `AstFactoryTest` for example?

Comment (tomekb234) [reply app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt:319]: A test to check if a correct AST is produced (in `AstFactoryTest`) would also be welcome, but what I mean is a simple integration test which checks that the loop actually works when running the program (in the `tests/` folder in the project root).

Comment (woocash2) [reply app/src/test/kotlin/compiler/e2e/CorrectProgramsE2eTest.kt:319]: Will add tomorrow, after I decode how to do it.

File: extras/vim/syntax/brzeczyk.vim [hunk 5]
```diff
@@ -23,7 +23,7 @@ syn keyword nnKeywordBoolLiteral fałsz prawda
 syn keyword nnKeywordOtherLiteral nic
 syn keyword nnKeywordCondition jeśli wpp
 syn match   nnKeywordElif '\<zaś gdy\>'
-syn keyword nnKeywordLoop dopóki otrzymując od
+syn keyword nnKeywordLoop dopóki otrzymując od dla
```
Comment (JHodorHodor) [inline extras/vim/syntax/brzeczyk.vim:26]: `wewnątrz`?

Comment (woocash2) [reply extras/vim/syntax/brzeczyk.vim:26]: Yeah

File: app/src/main/kotlin/compiler/syntax/AstFactory.kt [hunk 6]
```diff
@@ -106,6 +106,63 @@ class AstFactory(private val diagnostics: Diagnostics) {
         return rotateExpressionLeft(newTopLevelNode)
     }
 
+    fun makeArrayLoop(parseTree: ParseTree<Symbol>): Statement {
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/syntax/AstFactory.kt:109]: I meant for this function to take AST nodes as parameters.

Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/syntax/AstFactory.kt:109]: I think that this is also fine. I will have to refactor it a little bit merging anyway.

**Review by tomekb234 — APPROVED** [review]

File: app/src/main/kotlin/compiler/syntax/AstFactory.kt (hunk 6 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/syntax/AstFactory.kt:109]: Which ast nodes? `arrayExpression`, `receivingVariable` and `bodyBlock`?

Comment (woocash2) [reply app/src/main/kotlin/compiler/syntax/AstFactory.kt:109]: let's go

**Pull request merged**